<center><h5> Data Science For Suppy Chain & Operations </h5></center>
<center><h3> Etude de cas d'évaluation </h3></center>
<center><h1> Clustering de magasins </h1></center>

*Remplir ici votre compréhension du contexte et des objectifs*

# I. Chargement et création du jeu de données

In [8]:
import pandas as pd
from datetime import datetime
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns 
%matplotlib inline

In [9]:
#Ceci permet d'afficher 
pd.options.display.max_columns = 100
pd.options.display.max_rows  = 100

In [88]:
mag = pd.read_csv("magasins.csv", encoding = "latin", sep = ";", decimal = ',') #Le paramètre decimal définit comment est représenter la virgule des chiffres décimaux
ref = pd.read_csv("references_ete.csv", encoding = "latin", sep = ";")
ventes = pd.read_csv("ventes.csv", encoding = "latin", sep = ";")

## I.1. Exploration, nettoyage et préparation des données

Nous allons procéder à l'exploration des différents fichiers fournis et au nettoyage de chacun des fichiers

### I.1.1. Exploration des fichiers 

In [11]:
mag.head()


Réseau    Pays magasin      Magasin  Code magasin Type boutique  \
0  RETAIL EUROPE  UNITED KINGDOM   Magasin_99           286            GM   
1  RETAIL EUROPE          FRANCE   Magasin_49           381            GM   
2  RETAIL EUROPE          FRANCE   Magasin_39            41            GM   
3  RETAIL EUROPE       ALLEMAGNE  Magasin_193           303      BOUTIQUE   
4  RETAIL EUROPE          FRANCE  Magasin_190            59            GM   

   Surface magasin (*)   Climat Environnement concurrentiel Traffic Clientèle  \
0                 55.0    FROID                     PREMIUM     LOW    LOCALE   
1                 22.0    CHAUD                         NaN     NaN       NaN   
2                 75.0  TEMPERE                     PREMIUM     LOW    LOCALE   
3                139.0  TEMPERE                        LUXE     LOW    LOCALE   
4                 35.0  TEMPERE                     PREMIUM    HIGH    LOCALE   

      Saisonnalité                 Géographie Mall Comportement d'achat  \
0  PAS SAISONNIERE        CENTRE PETITE VILLE  OUI    CLIENT FULL PRICE   
1              NaN                        NaN  OUI    CLIENT FULL PRICE   
2  PAS SAISONNIERE        CENTRE PETITE VILLE  OUI      CLIENT DISCOUNT   
3  PAS SAISONNIERE  CENTRE VILLE GRANDE VILLE  NaN    CLIENT FULL PRICE   
4  PAS SAISONNIERE  CENTRE VILLE GRANDE VILLE  OUI      CLIENT DISCOUNT   

   Panier moyen € HT (Taux budget)  Indice de vente  
0                       267.438950         1.297333  
1                       207.102617         1.132231  
2                       195.375729         1.198321  
3                       320.788863         1.780049  
4                       171.843728         1.268786

In [28]:
print(mag.shape)
print(ref.shape)
print(ventes.shape)

(207, 16)
(357, 14)
(36726, 11)


In [91]:
ref.head()

Libellé rayon Libellé thème   Libellé famille Libellé sous famille  \
0     PAP FEMME   CHAINETRAME  BERMUDAS - SHORT                COTON   
1     PAP FEMME   CHAINETRAME  BERMUDAS - SHORT                COTON   
2     PAP FEMME   CHAINETRAME  BERMUDAS - SHORT                COTON   
3     PAP FEMME   CHAINETRAME  BERMUDAS - SHORT                COTON   
4     PAP FEMME   CHAINETRAME  BERMUDAS - SHORT              VISCOSE   

       Produit Libellé couleurs  PVB SPRING / SUMMER Puissance produit  \
0  Référence_0           LICHEN  150          SUMMER              BEST   
1  Référence_1            SABLE  150          SUMMER              BEST   
2  Référence_2             NOIR  150          SUMMER              SLOW   
3  Référence_3             BLEU  165          SUMMER            MEDIUM   
4  Référence_4       BUBBLE GUM  170          SUMMER            MEDIUM   

  Champs UVS Statut de collection Prix (selon famille) Iconique Type saison  \
0                      SAISONNIER               MEDIUM      NON      SUMMER   
1                      SAISONNIER               MEDIUM      NON      SUMMER   
2                      SAISONNIER               MEDIUM      NON      SUMMER   
3                      SAISONNIER               MEDIUM      NON      SUMMER   
4                      SAISONNIER               MEDIUM      NON      SUMMER   

  Segment  
0  CASUAL  
1  CASUAL  
2  CASUAL  
3  CASUAL  
4  CASUAL

In [13]:
ventes.head()

Magasin Référence produit Couleur produit        Modèle_Coloris  \
0  Magasin_127      Référence_14            BRUT      Référence_14BRUT   
1  Magasin_198     Référence_133         BLOSSOM  Référence_133BLOSSOM   
2   Magasin_10     Référence_279          MARINE   Référence_279MARINE   
3   Magasin_65      Référence_87            KAKI      Référence_87KAKI   
4  Magasin_100      Référence_28         FICELLE   Référence_28FICELLE   

                          SKUSTORE Collection produit Classification famille  \
0      Référence_14BRUTMagasin_127     E18 - ETE 2018               BLOUSONS   
1  Référence_133BLOSSOMMagasin_198     E18 - ETE 2018                   PULL   
2    Référence_279MARINEMagasin_10     E18 - ETE 2018                  VESTE   
3       Référence_87KAKIMagasin_65     E18 - ETE 2018                   JUPE   
4   Référence_28FICELLEMagasin_100     E18 - ETE 2018               CARDIGAN   

  Classification rayon Classification sous famille Statut Collection Produit  \
0            PAP FEMME                       COTON                SAISONNIER   
1            PAP FEMME                       COTON                SAISONNIER   
2            PAP FEMME                   POLYESTER                SAISONNIER   
3            PAP FEMME                 LIN / COTON                SAISONNIER   
4            PAP FEMME                    CASHMERE                SAISONNIER   

   Ventes17_19  
0           23  
1            4  
2            3  
3            1  
4            2

#### Fichier magasin

*Insérez vos conclusions ici*
- Le fichier **stores** est la liste des magasins que nous allons classifier. Ici, nous pouvons voir des informations sur chaque magasin. Il est lié au fichier des ventes par la colonne de Magasin.

Quel est le nombre de magasins présent dans le fichier ? De quelles données disposons-nous ?

mag (207, 16)
ref (357, 14)
ventes (36726, 11)

- Il y a 207 magasins avec 16 variables comme la localisation, le type, s'il est dans un centre commercial, le budget, la surface et l'indice de vente.

Quelques analyses à titre d'exemple...
- La variable Réseau peut être supprimée
- Nous pouvons uniquement garder le champ "Magasin" comme ID pour faire des jointures avec les autres tables, le champ code magasin peut être supprimé
- Il y a des données manquantes sur certains champs : 
    - Dans la colonne Mall, les valeurs manquantes signifient que les magasins ne sont pas dans des Malls
    - Dans d'autres colonnes catégoriques (Environnement concurrentiel, Traffic, Géographie, etc.) certains champs sont manquants. Il parait que l'information n'a pas été renseignée. Deux options s'offrent à nous : 
        - Supprimer ces colonnes 
        - Les compléter avec la valeur la plus présente
        - ...

        

In [82]:
print(mag.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207 entries, 0 to 206
Data columns (total 16 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Réseau                           207 non-null    object 
 1   Pays magasin                     207 non-null    object 
 2   Magasin                          207 non-null    object 
 3   Code magasin                     207 non-null    int64  
 4   Type boutique                    207 non-null    object 
 5   Surface magasin (*)              207 non-null    float64
 6   Climat                           207 non-null    object 
 7   Environnement concurrentiel      158 non-null    object 
 8   Traffic                          160 non-null    object 
 9   Clientèle                        160 non-null    object 
 10  Saisonnalité                     160 non-null    object 
 11  Géographie                       160 non-null    object 
 12  Mall                  

Afficher le nombre unique de "Magasin"


In [58]:
seriescountmagasin = pd.Series(mag["Magasin"].unique())
seriescountventemagasin = pd.Series(ventes["Magasin"].unique())

In [63]:
print("le nombre de magasins unique dans le fichier mag:", seriescountmagasin.value_counts().sum())
print("le nombre de magasins unique dans le fichier ventes:", seriescountventemagasin.value_counts().sum())

le nombre de magasins unique dans le fichier mag: 207
le nombre de magasins unique dans le fichier ventes: 168


In [65]:
categorical_variables = ['Réseau', 'Pays magasin','Type boutique',
       'Climat', 'Environnement concurrentiel','Traffic', 'Clientèle', 'Saisonnalité', 
        'Géographie', 'Mall', 'Comportement d\'achat']

In [69]:
numerical_variables = ['Surface magasin (*)', "Panier moyen \x80 HT (Taux budget)", "Indice de vente"]

In [120]:
ventes.head()

Magasin Référence produit Couleur produit        Modèle_Coloris  \
0  Magasin_127      Référence_14            BRUT      Référence_14BRUT   
1  Magasin_198     Référence_133         BLOSSOM  Référence_133BLOSSOM   
2   Magasin_10     Référence_279          MARINE   Référence_279MARINE   
3   Magasin_65      Référence_87            KAKI      Référence_87KAKI   
4  Magasin_100      Référence_28         FICELLE   Référence_28FICELLE   

                          SKUSTORE Classification famille  \
0      Référence_14BRUTMagasin_127               BLOUSONS   
1  Référence_133BLOSSOMMagasin_198                   PULL   
2    Référence_279MARINEMagasin_10                  VESTE   
3       Référence_87KAKIMagasin_65                   JUPE   
4   Référence_28FICELLEMagasin_100               CARDIGAN   

  Classification rayon Classification sous famille Statut Collection Produit  \
0            PAP FEMME                       COTON                SAISONNIER   
1            PAP FEMME                       COTON                SAISONNIER   
2            PAP FEMME                   POLYESTER                SAISONNIER   
3            PAP FEMME                 LIN / COTON                SAISONNIER   
4            PAP FEMME                    CASHMERE                SAISONNIER   

   Ventes17_19  
0           23  
1            4  
2            3  
3            1  
4            2

Afficher un comptage différentes modalités des variables catégoriques. Qu'en déduisez-vous ?

In [68]:
for cat in categorical_variables: 
    print("Variable : {}".format(cat))
    print(mag[cat].unique())
    print()

Variable : Réseau
['RETAIL EUROPE']

Variable : Pays magasin
['UNITED KINGDOM' 'FRANCE' 'ALLEMAGNE' 'PORTUGAL' 'ESPAGNE' 'ITALIE'
 'SUISSE' 'BELGIQUE' 'TURQUIE' 'SUEDE' 'NORVEGE' 'Danemark' 'PAYS-BAS'
 'GRECE' 'LUXEMBOURG' 'IRLANDE']

Variable : Type boutique
['GM' 'BOUTIQUE']

Variable : Climat
['FROID' 'CHAUD' 'TEMPERE']

Variable : Environnement concurrentiel
['PREMIUM' nan 'LUXE' 'MASS' 'PREMIUM / MASS']

Variable : Traffic
['LOW' nan 'HIGH' 'MEDIUM']

Variable : Clientèle
['LOCALE' nan 'TOURISTE']

Variable : Saisonnalité
['PAS SAISONNIERE' nan 'Saisonnière littoral' 'Saisonnière Montagne'
 'Saisonnière été']

Variable : Géographie
['CENTRE PETITE VILLE' nan 'CENTRE VILLE GRANDE VILLE'
 'centre ville moyen' 'PERIPHERIE' 'STATION SKI']

Variable : Mall
['OUI' nan]

Variable : Comportement d'achat
['CLIENT FULL PRICE' 'CLIENT DISCOUNT']



Afficher quelques statistiques sur les variables numériques

In [72]:
for num in numerical_variables : 
    print("Variable : {}".format(num))
    print(mag[num].describe())
    print()

Variable : Surface magasin (*)
count    207.000000
mean      80.697101
std       76.910501
min        0.000000
25%       35.000000
50%       59.500000
75%      105.000000
max      720.950000
Name: Surface magasin (*), dtype: float64

Variable : Panier moyen € HT (Taux budget)
count    205.000000
mean     248.167983
std       59.561461
min      125.165480
25%      212.074866
50%      237.550886
75%      276.948478
max      513.703718
Name: Panier moyen € HT (Taux budget), dtype: float64

Variable : Indice de vente
count    205.000000
mean       1.400662
std        0.277109
min        0.000000
25%        1.213813
50%        1.371019
75%        1.555035
max        2.177273
Name: Indice de vente, dtype: float64



Données manquantes : 

Afficher les données manquantes pour chaque variable.
Quelles données peut-on compléter ? Quels choix faites vous en terme de gestion des données manquantes et pourquoi ?

In [ ]:
#Afficher les données manquantes ici

In [83]:
mag.isnull().sum()

Réseau                              0
Pays magasin                        0
Magasin                             0
Code magasin                        0
Type boutique                       0
Surface magasin (*)                 0
Climat                              0
Environnement concurrentiel        49
Traffic                            47
Clientèle                          47
Saisonnalité                       47
Géographie                         47
Mall                               94
Comportement d'achat                0
Panier moyen € HT (Taux budget)     2
Indice de vente                     2
dtype: int64

In [74]:
mag["Pays magasin"].value_counts()

FRANCE            102
ESPAGNE            27
UNITED KINGDOM     17
SUISSE             14
ALLEMAGNE          11
ITALIE              9
PORTUGAL            6
PAYS-BAS            5
BELGIQUE            5
Danemark            3
GRECE               3
LUXEMBOURG          1
NORVEGE             1
IRLANDE             1
SUEDE               1
TURQUIE             1
Name: Pays magasin, dtype: int64

In [75]:
mag["Mall"] = mag["Mall"].fillna("NON")

Pour des raisons de simplicité, nous enlevons les variables contenant plus de 20 % de données manquantes. Vous êtes bien entendu libres de tester différentes imputations de données, comme imputer avec la valeur la plus présente par exemple.

Nous pouvons également certaines colonnes jugées inutiles (Réseau, Code Magasin). La colonne Pays Magasin peut être retravaillée si vous le souhaitez. Nous l'enlevons ici.

In [89]:
mag = mag.drop(["Réseau", "Pays magasin", "Code magasin", 
                'Environnement concurrentiel', 'Traffic', 'Clientèle', 'Saisonnalité',
       'Géographie'], axis = 1)

In [97]:
mag.head()

Magasin Type boutique  Surface magasin (*)   Climat Mall  \
0   Magasin_99            GM                 55.0    FROID  OUI   
1   Magasin_49            GM                 22.0    CHAUD  OUI   
2   Magasin_39            GM                 75.0  TEMPERE  OUI   
3  Magasin_193      BOUTIQUE                139.0  TEMPERE  NaN   
4  Magasin_190            GM                 35.0  TEMPERE  OUI   

  Comportement d'achat  Panier moyen € HT (Taux budget)  Indice de vente  
0    CLIENT FULL PRICE                       267.438950         1.297333  
1    CLIENT FULL PRICE                       207.102617         1.132231  
2      CLIENT DISCOUNT                       195.375729         1.198321  
3    CLIENT FULL PRICE                       320.788863         1.780049  
4      CLIENT DISCOUNT                       171.843728         1.268786

#### Fichier ventes

*Insérer vos conclusions ici.*
- Le fichier des ventes contient toutes les ventes des magasins. Le variable clé avec fichier magasin est Magasin.
- Il n'y a pas de variables null dans le fichier.

In [96]:
ventes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36726 entries, 0 to 36725
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   Magasin                      36726 non-null  object
 1   Référence produit            36726 non-null  object
 2   Couleur produit              36726 non-null  object
 3   Modèle_Coloris               36726 non-null  object
 4   SKUSTORE                     36726 non-null  object
 5   Collection produit           36726 non-null  object
 6   Classification famille       36726 non-null  object
 7   Classification rayon         36726 non-null  object
 8   Classification sous famille  36726 non-null  object
 9   Statut Collection Produit    36726 non-null  object
 10  Ventes17_19                  36726 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 3.1+ MB


Combien y'a-t-il de magasins dans ce fichier, combien de y'a-t-il de produits ? Comparer le nombre de magasins ici avec le fichier magasins.

In [99]:
# Explorer le dataset
seriescountmagasin = pd.Series(mag["Magasin"].unique())
seriescountventemagasin = pd.Series(ventes["Magasin"].unique())
print("le nombre de magasins unique dans le fichier mag:", seriescountmagasin.value_counts().sum())
print("le nombre de magasins unique dans le fichier ventes:", seriescountventemagasin.value_counts().sum())

seriescountventeproduit = pd.Series(ventes["Référence produit"].unique())
print("le nombre de produits unique dans le fichier ventes:", seriescountventeproduit.value_counts().sum())

le nombre de magasins unique dans le fichier mag: 207
le nombre de magasins unique dans le fichier ventes: 168
le nombre de produits unique dans le fichier ventes: 284


In [111]:
categorical_values_ventes = ["Collection produit", 'Classification famille',
       'Classification rayon', 'Classification sous famille',
       'Statut Collection Produit']
numerical_values_ventes = ['Ventes17_19']

Afficher un comptage différentes modalités des variables catégoriques. Qu'en déduisez-vous ?

In [112]:
for cat in categorical_values_ventes :
    print("Variable : {}".format(cat))
    print(ventes[cat].unique())
    print()

Variable : Collection produit
['E18 - ETE 2018']

Variable : Classification famille
['BLOUSONS' 'PULL' 'VESTE' 'JUPE' 'CARDIGAN' 'TUNIQUE ML' 'DEBARDEUR'
 'TEE SHIRT' 'TUNISIEN' 'ROBE' 'PANTALON' 'SWEAT' 'BERMUDAS - SHORT'
 'JEANS' 'CHEMISE' 'CARACO' 'CUIR-FOURRURE' 'TOP' 'PARKAS']

Variable : Classification rayon
['PAP FEMME']

Variable : Classification sous famille
['COTON' 'POLYESTER' 'LIN / COTON' 'CASHMERE' 'LAINE' 'MODAL'
 'CACHEMIRE PLUME' 'LIN' 'MERINOS' 'ACRYLIC MELANGE' 'COTON / VISCOSE'
 'LAINE / CASH' 'VISCOSE' 'ACETATE' 'ALPAGA' 'SOIE' 'COTON / SOIE' 'CUIR'
 'MOHAIR']

Variable : Statut Collection Produit
['SAISONNIER' 'PERMANENT' 'NINI PERMANENT']



In [113]:
ventes[numerical_values_ventes[0]].describe()

count    36726.000000
mean         6.402630
std          5.893397
min         -1.000000
25%          3.000000
50%          5.000000
75%          8.000000
max        188.000000
Name: Ventes17_19, dtype: float64

In [114]:
ventes.isnull().sum()

Magasin                        0
Référence produit              0
Couleur produit                0
Modèle_Coloris                 0
SKUSTORE                       0
Collection produit             0
Classification famille         0
Classification rayon           0
Classification sous famille    0
Statut Collection Produit      0
Ventes17_19                    0
dtype: int64

In [116]:
ventes["Collection produit"].unique()

array(['E18 - ETE 2018'], dtype=object)

In [117]:
ventes = ventes.drop(["Collection produit"], axis = 1)

#### Fichier de références

*Insérez vos conclusions ici.*

- Le fichier de référence contient des informations sur le type de produits vendus dans les magasins.


- Le fichier de références contient des informations sur 288 produits, et 357 lignes contenant avec le détail du coloris.
- Il convient de faire une jointure entre le fichier des références et des ventes. Pour ceci, il faut créer la clé de jointure :
    - Pour la table ventes, il s'agit du champ Modèle_Coloris, qui est une concaténation des champs "Référence produit" et "Couleur produit"
    - Pour la table références, il faut créer la clé de jointure en concaténant le champ "Produit" et le champ "Libellé couleurs"

In [118]:
# Exploration de jeu de données
print(ref.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 357 entries, 0 to 356
Data columns (total 14 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   Libellé rayon                    357 non-null    object
 1   Libellé thème                    357 non-null    object
 2   Libellé famille                  357 non-null    object
 3   Libellé sous famille             333 non-null    object
 4   Produit                          357 non-null    object
 5   Libellé couleurs                 357 non-null    object
 6   PVB                              357 non-null    int64 
 7   SPRING / SUMMER                  335 non-null    object
 8   Puissance produit                356 non-null    object
 9   Champs UVS Statut de collection  357 non-null    object
 10  Prix (selon famille)             357 non-null    object
 11  Iconique                         357 non-null    object
 12  Type saison                      357

In [119]:
seriescountrefprod = pd.Series(ref["Produit"].unique())
print("le nombre de produit unique dans le fichier ref:", seriescountrefprod.value_counts().sum())


le nombre de produit unique dans le fichier ref: 288


In [131]:
# Exploration des variables catégoriques et numériques
categorical_variables_ref = ['Libellé rayon', 'Libellé thème', 'Libellé famille', 'Libellé sous famille', 'Produit',	
       'Libellé couleurs', 'SPRING / SUMMER','Puissance produit', 'Champs UVS Statut de collection',
       'Prix (selon famille)', 'Iconique', 'Type saison', 'Segment']
numerical_variables_ref = ['PVB']

In [138]:
for cat in categorical_variables_ref: 
    print("Variable : {}".format(cat))
    print(ref[cat].unique())
    print()

Variable : Libellé rayon
['PAP FEMME']

Variable : Libellé thème
['CHAINETRAME' 'MAILLE' 'JERSEY']

Variable : Libellé famille
['BERMUDAS - SHORT' 'BLOUSONS' 'CARACO' 'CARDIGAN' 'CHEMISE'
 'CUIR-FOURRURE' 'DEBARDEUR' 'JEANS' 'JUPE' 'PANTALON' 'PARKAS' 'PULL'
 'ROBE' 'SWEAT' 'TEE SHIRT' 'TOP' 'TUNIQUE ML' 'TUNISIEN' 'VESTE']

Variable : Libellé sous famille
['COTON' 'VISCOSE' 'LAINE' 'SOIE' 'MERINOS' 'CASHMERE' 'ALPAGA'
 'LAINE / CASH' 'CUIR' 'COTON / VISCOSE' 'MODAL' 'COTON / SOIE'
 'LIN / COTON' 'POLYESTER' 'ACETATE' 'ACRYLIC MELANGE' 'LIN'
 'CACHEMIRE PLUME']

Variable : Produit
['Référence_0' 'Référence_1' 'Référence_2' 'Référence_3' 'Référence_4'
 'Référence_5' 'Référence_6' 'Référence_7' 'Référence_8' 'Référence_9'
 'Référence_10' 'Référence_11' 'Référence_12' 'Référence_13'
 'Référence_14' 'Référence_15' 'Référence_16' 'Référence_17'
 'Référence_18' 'Référence_19' 'Référence_20' 'Référence_21'
 'Référence_22' 'Référence_23' 'Référence_24' 'Référence_25'
 'Référence_26' 'Référence

In [127]:
for num in numerical_variables_ref : 
    print("Variable : {}".format(num))
    print(ref[num].describe())
    print()

Variable : PVB
count    357.00000
mean     256.24930
std      165.36061
min       45.00000
25%      120.00000
50%      225.00000
75%      340.00000
max      990.00000
Name: PVB, dtype: float64



In [128]:
ref.isnull().sum()

Libellé rayon                       0
Libellé thème                       0
Libellé famille                     0
Libellé sous famille               24
Produit                             0
Libellé couleurs                    0
PVB                                 0
SPRING / SUMMER                    22
Puissance produit                   1
Champs UVS Statut de collection     0
Prix (selon famille)                0
Iconique                            0
Type saison                         0
Segment                             0
dtype: int64

In [134]:
# On rempli les valeurs nan avec des valeurs random de chaque colonne.
import random
ref['Libellé sous famille'].fillna(random.choice(ref['Libellé sous famille'][ref['Libellé sous famille'].notna()]), inplace=True)

In [136]:
ref['SPRING / SUMMER'].fillna(random.choice(ref['SPRING / SUMMER'][ref['SPRING / SUMMER'].notna()]), inplace=True)

In [139]:
# Comme il n'y a qu'une seule valeur nan dans la colonne "Puissance produit" nous la remplissons avec MEDIUM.
ref['Puissance produit'] = ref['Puissance produit'].fillna('MEDIUM')

In [140]:
ref.isnull().sum()

Libellé rayon                      0
Libellé thème                      0
Libellé famille                    0
Libellé sous famille               0
Produit                            0
Libellé couleurs                   0
PVB                                0
SPRING / SUMMER                    0
Puissance produit                  0
Champs UVS Statut de collection    0
Prix (selon famille)               0
Iconique                           0
Type saison                        0
Segment                            0
dtype: int64

In [141]:
#Création  de la clé de jointure
ref["Modèle_Coloris"] = ref["Produit"] + ref["Libellé couleurs"]

### I.1.2. Jointure entre le fichier de ventes et de références

Nous effectuons une jointure "inner" entre les ventes et les références. L'objectif est de ne garder que les informations contenues dans les deux fichiers.

Nous calculons ensuite le montant total pour chaque référence produit et chaque magasin.

In [143]:
ventes_ref = ventes.merge(ref, how = 'inner', on = "Modèle_Coloris")

In [144]:
# Calculer le Montant Total
ventes_ref["Montant total"] = ventes_ref["Ventes17_19"] * ventes_ref["PVB"]

In [146]:
ventes_ref.head()

Magasin Référence produit Couleur produit    Modèle_Coloris  \
0  Magasin_127      Référence_14            BRUT  Référence_14BRUT   
1  Magasin_193      Référence_14            BRUT  Référence_14BRUT   
2  Magasin_154      Référence_14            BRUT  Référence_14BRUT   
3   Magasin_81      Référence_14            BRUT  Référence_14BRUT   
4   Magasin_84      Référence_14            BRUT  Référence_14BRUT   

                      SKUSTORE Classification famille Classification rayon  \
0  Référence_14BRUTMagasin_127               BLOUSONS            PAP FEMME   
1  Référence_14BRUTMagasin_193               BLOUSONS            PAP FEMME   
2  Référence_14BRUTMagasin_154               BLOUSONS            PAP FEMME   
3   Référence_14BRUTMagasin_81               BLOUSONS            PAP FEMME   
4   Référence_14BRUTMagasin_84               BLOUSONS            PAP FEMME   

  Classification sous famille Statut Collection Produit  Ventes17_19  \
0                       COTON                SAISONNIER           23   
1                       COTON                SAISONNIER            4   
2                       COTON                SAISONNIER           16   
3                       COTON                SAISONNIER            8   
4                       COTON                SAISONNIER            7   

  Libellé rayon Libellé thème Libellé famille Libellé sous famille  \
0     PAP FEMME   CHAINETRAME        BLOUSONS                COTON   
1     PAP FEMME   CHAINETRAME        BLOUSONS                COTON   
2     PAP FEMME   CHAINETRAME        BLOUSONS                COTON   
3     PAP FEMME   CHAINETRAME        BLOUSONS                COTON   
4     PAP FEMME   CHAINETRAME        BLOUSONS                COTON   

        Produit Libellé couleurs  PVB SPRING / SUMMER Puissance produit  \
0  Référence_14             BRUT  380          SPRING            MEDIUM   
1  Référence_14             BRUT  380          SPRING            MEDIUM   
2  Référence_14             BRUT  380          SPRING            MEDIUM   
3  Référence_14             BRUT  380          SPRING            MEDIUM   
4  Référence_14             BRUT  380          SPRING            MEDIUM   

  Champs UVS Statut de collection Prix (selon famille) Iconique Type saison  \
0                      SAISONNIER                 HIGH      OUI  MID-SEASON   
1                      SAISONNIER                 HIGH      OUI  MID-SEASON   
2                      SAISONNIER                 HIGH      OUI  MID-SEASON   
3                      SAISONNIER                 HIGH      OUI  MID-SEASON   
4                      SAISONNIER                 HIGH      OUI  MID-SEASON   

  Segment  Montant total  
0  CASUAL           8740  
1  CASUAL           1520  
2  CASUAL           6080  
3  CASUAL           3040  
4  CASUAL           2660

### I.1.3. Création de variables à la maille du magasin sur le fichier joint des ventes et des références

Etant donné que nous souhaitons faire un clustering de magasins, il convient de calculer de nouvelles features à partir du dataset "ventes_ref" à la maille du magasin.
La fonctionnalité groupby permet de répondre à ce besoin.


**Calcul du montant total et du nombre de ventes total par magasin**

In [154]:
mag_montants_ventes = ventes_ref.groupby("Magasin")[["Montant total", "Ventes17_19"]].sum()

In [157]:
mag_montants_ventes.head()

Montant total  Ventes17_19
Magasin                                
Magasin_1           512234         2283
Magasin_10          181469          856
Magasin_100         593470         2546
Magasin_101         458868         2169
Magasin_102         163359          793

Calculez le prix de vente moyen

In [159]:
mag_montants_ventes["Prix de vente moyen"] = mag_montants_ventes["Montant total"] / mag_montants_ventes["Ventes17_19"]

In [160]:
mag_montants_ventes.head()

Montant total  Ventes17_19  Prix de vente moyen
Magasin                                                     
Magasin_1           512234         2283           224.368813
Magasin_10          181469          856           211.996495
Magasin_100         593470         2546           233.098979
Magasin_101         458868         2169           211.557400
Magasin_102         163359          793           206.001261

**Calcul du pourcentage de ventes des produits iconiques par magasin** 

In [161]:
ventes_ref["Iconique"].value_counts(dropna = False)

NON    22363
OUI    14268
Name: Iconique, dtype: int64

In [ ]:
#On remplace les champs Oui et Non par "Iconique" et "Non_iconique" pour pouvoir les reconnaitre plus tard

In [162]:
ventes_ref["Iconique_transformed"] = ventes_ref["Iconique"].map({"OUI":"Iconique", "NON": "Non_iconique"})

In [163]:
iconique_sum = ventes_ref.groupby(["Magasin", "Iconique_transformed"])['Ventes17_19'].sum().unstack()

In [164]:
iconique_sum.head()

Iconique_transformed  Iconique  Non_iconique
Magasin                                     
Magasin_1                931.0        1352.0
Magasin_10               339.0         517.0
Magasin_100             1009.0        1537.0
Magasin_101              877.0        1292.0
Magasin_102              345.0         448.0

Nous remarquons certaines valeurs manquantes. Cela veut dire que dans certains magasins (en l'occurence, les magasins 146 et 89), aucun produit iconique n'a été vendu. Nous allons donc imputer les valeurs manquantes à 0.

In [165]:
iconique_sum[iconique_sum.Iconique.isnull()]

Iconique_transformed  Iconique  Non_iconique
Magasin                                     
Magasin_118                NaN           1.0
Magasin_61                 NaN           1.0

In [166]:
iconique_sum = iconique_sum.fillna(0)

In [167]:
iconique_prc = iconique_sum.div(iconique_sum.sum(axis = 1), axis = 0)

In [168]:
iconique_prc.head()

Iconique_transformed  Iconique  Non_iconique
Magasin                                     
Magasin_1             0.407797      0.592203
Magasin_10            0.396028      0.603972
Magasin_100           0.396308      0.603692
Magasin_101           0.404334      0.595666
Magasin_102           0.435057      0.564943

**Faire le même traitement pour d'autres colonnes (nous proposon ces colonnes, mais il est possible d'en rajouter d'autres si vous les trouvez pertinentes) :**
- Prix (selon famille)
- Type saison
- Segment
- Libellé thème
- Statut de collection

Il peut être pertinent de créer une boucle for et de stocker les datasets temporaires dans une liste, puis de les concaténer

In [169]:
#On remplace les champs Oui et Non par "Iconique" et "Non_iconique" pour pouvoir les reconnaitre plus tard
ventes_ref["Iconique_transformed"] = ventes_ref["Iconique"].map({"OUI":"Iconique", "NON": "Non_iconique"})

#On rajoute le nom de la feature pour reconnaitre les catégories
ventes_ref["Prix (selon famille)_transformed"] = "Famille prix_"+ventes_ref["Prix (selon famille)"]

#Même chose pour le type de saison
ventes_ref["Type saison_transformed"] = "Type saison_"+ventes_ref["Type saison"]

#Même chose pour le Segment
ventes_ref["Segment_transformed"] = "Segment_"+ventes_ref["Type saison"]

#Même chose pour le Libellé thème
ventes_ref["Libellé thème_transformed"] = "Libellé thème_"+ventes_ref["Libellé thème"]

#Même chose pour le Libellé thème
ventes_ref["Statut Collection Produit_transformed"] = "Statut Collection Produit_"+ventes_ref["Libellé thème"]

In [171]:
ventes_ref.head()

Magasin Référence produit Couleur produit    Modèle_Coloris  \
0  Magasin_127      Référence_14            BRUT  Référence_14BRUT   
1  Magasin_193      Référence_14            BRUT  Référence_14BRUT   
2  Magasin_154      Référence_14            BRUT  Référence_14BRUT   
3   Magasin_81      Référence_14            BRUT  Référence_14BRUT   
4   Magasin_84      Référence_14            BRUT  Référence_14BRUT   

                      SKUSTORE Classification famille Classification rayon  \
0  Référence_14BRUTMagasin_127               BLOUSONS            PAP FEMME   
1  Référence_14BRUTMagasin_193               BLOUSONS            PAP FEMME   
2  Référence_14BRUTMagasin_154               BLOUSONS            PAP FEMME   
3   Référence_14BRUTMagasin_81               BLOUSONS            PAP FEMME   
4   Référence_14BRUTMagasin_84               BLOUSONS            PAP FEMME   

  Classification sous famille Statut Collection Produit  Ventes17_19  \
0                       COTON                SAISONNIER           23   
1                       COTON                SAISONNIER            4   
2                       COTON                SAISONNIER           16   
3                       COTON                SAISONNIER            8   
4                       COTON                SAISONNIER            7   

  Libellé rayon Libellé thème Libellé famille Libellé sous famille  \
0     PAP FEMME   CHAINETRAME        BLOUSONS                COTON   
1     PAP FEMME   CHAINETRAME        BLOUSONS                COTON   
2     PAP FEMME   CHAINETRAME        BLOUSONS                COTON   
3     PAP FEMME   CHAINETRAME        BLOUSONS                COTON   
4     PAP FEMME   CHAINETRAME        BLOUSONS                COTON   

        Produit Libellé couleurs  PVB SPRING / SUMMER Puissance produit  \
0  Référence_14             BRUT  380          SPRING            MEDIUM   
1  Référence_14             BRUT  380          SPRING            MEDIUM   
2  Référence_14             BRUT  380          SPRING            MEDIUM   
3  Référence_14             BRUT  380          SPRING            MEDIUM   
4  Référence_14             BRUT  380          SPRING            MEDIUM   

  Champs UVS Statut de collection Prix (selon famille) Iconique Type saison  \
0                      SAISONNIER                 HIGH      OUI  MID-SEASON   
1                      SAISONNIER                 HIGH      OUI  MID-SEASON   
2                      SAISONNIER                 HIGH      OUI  MID-SEASON   
3                      SAISONNIER                 HIGH      OUI  MID-SEASON   
4                      SAISONNIER                 HIGH      OUI  MID-SEASON   

  Segment  Montant total Iconique_transformed  \
0  CASUAL           8740             Iconique   
1  CASUAL           1520             Iconique   
2  CASUAL           6080             Iconique   
3  CASUAL           3040             Iconique   
4  CASUAL           2660             Iconique   

  Prix (selon famille)_transformed Type saison_transformed  \
0                Famille prix_HIGH  Type saison_MID-SEASON   
1                Famille prix_HIGH  Type saison_MID-SEASON   
2                Famille prix_HIGH  Type saison_MID-SEASON   
3                Famille prix_HIGH  Type saison_MID-SEASON   
4                Famille prix_HIGH  Type saison_MID-SEASON   

  Segment_transformed  Libellé thème_transformed  \
0  Segment_MID-SEASON  Libellé thème_CHAINETRAME   
1  Segment_MID-SEASON  Libellé thème_CHAINETRAME   
2  Segment_MID-SEASON  Libellé thème_CHAINETRAME   
3  Segment_MID-SEASON  Libellé thème_CHAINETRAME   
4  Segment_MID-SEASON  Libellé thème_CHAINETRAME   

   Statut Collection Produit_transformed  
0  Statut Collection Produit_CHAINETRAME  
1  Statut Collection Produit_CHAINETRAME  
2  Statut Collection Produit_CHAINETRAME  
3  Statut Collection Produit_CHAINETRAME  
4  Statut Collection Produit_CHAINETRAME

In [172]:
ventes_ref["Prix (selon famille)"].value_counts(dropna = False)

MEDIUM    20714
LOW        9208
HIGH       6709
Name: Prix (selon famille), dtype: int64

In [173]:
prix_famille_sum = ventes_ref.groupby(["Magasin", "Prix (selon famille)_transformed"])['Ventes17_19'].sum().unstack()

In [175]:
prix_famille_sum.head()

Prix (selon famille)_transformed  Famille prix_HIGH  Famille prix_LOW  \
Magasin                                                                 
Magasin_1                                     439.0             558.0   
Magasin_10                                    121.0             251.0   
Magasin_100                                   523.0             574.0   
Magasin_101                                   338.0             591.0   
Magasin_102                                   107.0             186.0   

Prix (selon famille)_transformed  Famille prix_MEDIUM  
Magasin                                                
Magasin_1                                      1286.0  
Magasin_10                                      484.0  
Magasin_100                                    1449.0  
Magasin_101                                    1240.0  
Magasin_102                                     500.0

Réalisez les mêmes traitements que pour le champ "Iconique"

In [ ]:
# features_magasin = [iconique_prc,

**Nous vous encourageons à rajouter de nouvelles variables si elles vous paraissent pertinentes.**

### I.1.4. Concaténation avec le fichier des magasins

In [ ]:
mags_final = mag.merge(features_magasins, left_on = "Magasin", right_index = True )

In [ ]:
# nous définissons la colonne Magasin comme étant l'index de chaque ligne
mags_final = mags_final.set_index('Magasin')

In [ ]:
mags_df = mags_final[cols_num].copy()

# II. Analyse en composantes principales

In [ ]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler,MinMaxScaler
import numpy as np

Comme expliqué en cours, l'analyse en composantes principales (PCA) est une méthode factorielle de réduction de dimension pour l’exploration de données quantitatives. Elle permet de transformer des variables corrélées en nouvelles variables décorrélées, appelées composantes principales.

In [ ]:
# Réaliser une PCA Ici (pensez à tester avec des données normalisées)


In [ ]:
# Explained variance ratio des composantes principales (qu'on pourrait comparer avec celles des features originales...)


Représentez des magasins sur les composantes principales

Analysez les composantes principales comme vu en cours : Quelles sont les variables les plus corrélées avec les composantes principales ? 

# III. Clustering

In [ ]:
from sklearn.cluster import KMeans

## II.1. Méthode du coude pour définir le nombre de clusters

Il peut être intéressant de tester le clustering avec les données brutes, les données normalisées, et les données transformées avec l'ACP.

In [ ]:
#Méthode du coude

In [ ]:
# Entrainer un k-means ici


In [ ]:
# Ajoutez les labels des clusters au dataframe

### Caractéristisation des clusters :

Analysez chacun des clusters : Quelle est la distribution des variables ? Nous vous invitons à filtrer sur chacun des clusters et à établir des statistiques.

*Commencez avec des analyses statistiques simples*

Taille des clusters

In [ ]:
mags_df["Cluster"].value_counts()

In [ ]:
mags_df.groupby("Cluster").mean()

In [ ]:
mags_df.mean()

**Portez une attention particulière à l'analyse et l'interprétation des premiers résultats du clustering. Ces résultats pourront vous inspirer la création de nouvelles variables plus pertinentes, ou l'adoption d'une nouvelle démarche. Par exemple, si on trouve un cluster principal qui contient la majorité des magasins et d'autres clusters avec un très faible nombre de magasins, il peut être intéressant de les écarter et de refaire un clustering uniquement sur le gros cluster.**

**On tombe rarement sur la bonne réponse du premier coup, c'est une approche très itérative.**
